pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/AndFilterOperator.java

```diff
@@ -28,6 +28,7 @@
   private final List<BaseFilterOperator> _filterOperators;
 
   public AndFilterOperator(List<BaseFilterOperator> filterOperators) {
+    // NOTE: EmptyFilterOperator and MatchAllFilterOperator should not be passed into the AndFilterOperator
```

Comment:
Is that really necessary? Why can't we handle these operators if they are passed in more gracefully?

Reply:
+1

Reply:
We can handle them, but we should not pass them in per performance concern (actually we won't pass them in because re-order operators won't take these 2 operators). I just added this comment for clarification. Will add more details to the comment.

Reply:
Additional comments/context is great - thanks for adding that; however, it might be better to programatically enforce it via constructor checks (throw IllegalArg/State Exc for example) so the contract is not broken when code gets refactored/changed. In other places in this PR you have used asserts. These are weak checks as they can be disabled at the JVM level; would be good to consider throwing exceptions instead.